File: Native/GaiaLogClient/LogClient.cpp
#include "LogClient.hpp"

#include <iostream>

namespace Gaia::LogService
{
    /// Try to connect to the Redis server, and it will use a local file instead if failed.
    LogClient::LogClient(unsigned int port, const std::string &ip)
    {
        try
        {
            Connection = std::make_unique<sw::redis::Redis>("tcp://" + ip + ":" + std::to_string(port));
            if (Connection->publish("logs/record",
                                    LogRecorder::GenerateLogText("Log service client connected.",
                                    LogRecorder::Severity::Message, Author)) <1)
            {
                throw std::runtime_error("No log server detected on " + ip + ":" + std::to_string(port));
            }
        }catch (std::exception& error)
        {
            Connection.reset();
            Logger = std::make_unique<LogRecorder>();
            Logger->RecordError(error.what(), Author);
            Logger->RecordError("Failed to connect the Redis server on " + ip + ":" + std::to_string(port));
        }
    }

    /// Record a raw text into the log.
    void LogClient::RecordRawText(const std::string& text)
    {
        if (Connection)
        {
            Connection->publish("logs/record", text);
        }
        else if (Logger)
        {
            Logger->RecordRawText(text);
        }

        if (PrintToConsole)
        {
            std::cout << text << std::endl;
        }
    }

    /// Record a message log.
    void LogClient::RecordMessage(const std::string& text)
    {
        RecordRawText(LogRecorder::GenerateLogText(text, LogRecorder::Severity::Message, Author));
    }

    /// Record a milestone log.
    void LogClient::RecordMilestone(const std::string& text)
    {
        RecordRawText(LogRecorder::GenerateLogText(text, LogRecorder::Severity::Milestone, Author));
    }

    /// Record a warning log.
    void LogClient::RecordWarning(const std::string& text)
    {
        RecordRawText(LogRecorder::GenerateLogText(text, LogRecorder::Severity::Warning, Author));
    }

    /// Record an error log.
    void LogClient::RecordError(const std::string& text)
    {
        RecordRawText(LogRecorder::GenerateLogText(text, LogRecorder::Severity::Error, Author));
    }
}

File: GaiaLogClient/LogClient.cs
using System;
using System.Linq;
using System.Threading;
using StackExchange.Redis;

namespace Gaia.LogService
{
    /// <summary>
    /// Client for log service, providing log recording functions.
    /// If this client can not connect to the Redis server,
    /// it will use a local log file instead.
    /// </summary>
    public class LogClient
    {
        /// <summary>
        /// The subscriber to the Redis server to use the log service.
        /// </summary>
        private readonly ISubscriber Subscriber = null;
        /// <summary>
        /// If subscriber failed to connect to the Redis server, this local file log will be used instead.
        /// </summary>
        private readonly LogRecorder Logger = null;

        /// <summary>
        /// Whether print logs to the console or not.
        /// If it is true, logs will be sent to the log server and then printed to the local console.
        /// </summary>
        public bool PrintToConsole = false;

        /// <summary>
        /// The author of the logs sent from this client.
        /// </summary>
        public string Author = "Anonymous";
        
        /// <summary>
        /// Construct and try to connect to the Redis server.
        /// </summary>
        /// <param name="port">Port of the Redis server.</param>
        /// <param name="ip">IP address of the Redis server.</param>
        public LogClient(uint port = 6379, string ip = "127.0.0.1") : 
            this(ConnectionMultiplexer.Connect($"{ip}:{port.ToString()}"))
        { }

        /// <summary>
        /// Reuse the connection to a Redis server.
        /// </summary>
        /// <param name="connection">Connection to the Redis server.</param>
        public LogClient(IConnectionMultiplexer connection)
        {
            Subscriber = connection.GetSubscriber();
            if (Subscriber.Publish(
                "logs/record",
                LogRecorder.GenerateLogText(
                    "Log service client connected.",
                    LogRecorder.Severity.Message, Author)) < 1)
            {
                Logger = new LogRecorder {PrintToConsole = false};
                Subscriber = null;
            }
        }

        /// <summary>
        /// Use a local log file.
        /// </summary>
        public LogClient()
        {
            Logger = new LogRecorder {PrintToConsole = false};
        }
        
        /// <summary>
        /// Record a log.
        /// </summary>
        private void RecordRawText(string text)
        {
            if (Subscriber != null)
            {
                Subscriber.Publish("logs/record", text);
            }
            else if (Logger != null)
            {
                Logger.RecordRawText(text);
            }

            if (PrintToConsole)
            {
                Console.WriteLine(text);
            }
        }
        
        /// <summary>
        /// Record a milestone log.
        /// Milestone log represents important time points of a program. 
        /// </summary>
        /// <param name="text">Text of the log.</param>
        public void RecordMilestone(string text)
        {
            RecordRawText(LogRecorder.GenerateLogText(text, LogRecorder.Severity.Milestone, Author));
        }

        /// <summary>
        /// Record a message log.
        /// Message log represents simple output of a program.
        /// </summary>
        /// <param name="text">Text of the log.</param>
        public void RecordMessage(string text)
        {
            RecordRawText(LogRecorder.GenerateLogText(text, LogRecorder.Severity.Message, Author));
        }

        /// <summary>
        /// Record a warning log.
        /// Warning log represents important messages that should pay attention to.
        /// </summary>
        /// <param name="text">Text of the log.</param>
        public void RecordWarning(string text)
        {
            RecordRawText(LogRecorder.GenerateLogText(text, LogRecorder.Severity.Warning, Author));
        }

        /// <summary>
        /// Record an error log.
        /// Error log represents the abnormal situation of a program.
        /// </summary>
        /// <param name="text">Text of the log.</param>
        public void RecordError(string text)
        {
            RecordRawText(LogRecorder.GenerateLogText(text, LogRecorder.Severity.Error, Author));
        }
    }
}